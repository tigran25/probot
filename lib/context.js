/**
 * Helpers for extracting information from the webhook event, which can be
 * passed to GitHub API calls.
 *
 * @property {github} github - An authenticated GitHub API client
 * @property {payload} payload - The webhook event payload
 */
class Context {
  constructor(event, github) {
    Object.assign(this, event);
    this.github = github;
  }

  /**
   * Return the `owner` and `repo` params for making API requests against a
   * repository.
   *
   * @param {object} [object] - Params to be merged with the repo params.
   *
   * @example
   *
   * const params = context.repo({path: '.github/stale.yml'})
   * // Returns: {owner: 'username', repo: 'reponame', path: '.github/stale.yml'}
   *
   */
  repo(object) {
    const repo = this.payload.repository;

    return Object.assign({
      owner: repo.owner.login || repo.owner.name,
      repo: repo.name
    }, object);
  }

  /**
   * Return the `owner`, `repo`, and `number` params for making API requests
   * against an issue or pull request. The object passed in will be merged with
   * the repo params.
   *
   * @example
   *
   * const params = context.issue({body: 'Hello World!'})
   * // Returns: {owner: 'username', repo: 'reponame', number: 123, body: 'Hello World!'}
   *
   * @param {object} [object] - Params to be merged with the issue params.
   */
  issue(object) {
    const payload = this.payload;
    return Object.assign({
      number: (payload.issue || payload.pull_request || payload).number
    }, this.repo(), object);
  }

  /**
   * Returns a boolean if the actor on the event was a bot.
   * @type {boolean}
   */
  get isBot() {
    return this.payload.sender.type === 'Bot';
  }
}

module.exports = Context;
