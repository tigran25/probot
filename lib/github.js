const Bottleneck = require('bottleneck')
const Octokit = require('@octokit/rest')

/**
 * the [@octokit/rest Node.js module](https://github.com/octokit/rest.js),
 * which wraps the [GitHub API](https://developer.github.com/v3/) and allows
 * you to do almost anything programmatically that you can do through a web
 * browser.
 * @typedef github
 * @see {@link https://github.com/octokit/rest.js}
 */

const defaultCallback = response => response
async function paginate (octokit, responsePromise, callback = defaultCallback) {
  let collection = []
  let getNextPage = true
  let done = () => {
    getNextPage = false
  }
  let response = await responsePromise
  collection = collection.concat(await callback(response, done))
  // eslint-disable-next-line no-unmodified-loop-condition
  while (getNextPage && octokit.hasNextPage(response)) {
    response = await octokit.getNextPage(response)
    collection = collection.concat(await callback(response, done))
  }
  return collection
}

function EnhancedGitHubClient (options) {
  const octokit = Octokit(options)
  const limiter = options.limiter || new Bottleneck({ maxConcurrent: 1, minTime: 1000 })
  const logger = options.logger
  const noop = () => Promise.resolve()

  octokit.hook.before('request', limiter.schedule.bind(limiter, noop))
  octokit.hook.error('request', (error, options) => {
    const {method, url, headers, ...params} = options
    const msg = `GitHub request: ${method} ${url} - ${error.code} ${error.status}`
    logger.debug({params}, msg)
    throw error
  })
  octokit.hook.after('request', (result, options) => {
    const {method, url, headers, ...params} = options
    const msg = `GitHub request: ${method} ${url} - ${result.meta.status}`
    logger.debug({params}, msg)
  })
  octokit.paginate = paginate.bind(null, octokit)

  return octokit
}

module.exports = EnhancedGitHubClient
