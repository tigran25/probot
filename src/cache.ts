import cacheManager from 'cache-manager'

// The TypeScript definition for cache-manager does not export the Cache interface so we recreate it here
export interface Cache {
  wrap<T> (key: string, wrapper: (callback: (error: any, result: T) => void) => any, options: CacheConfig): Promise<any>
}
export interface CacheConfig {
  ttl: number
}

export function createDefaultCache (): Cache {
  return cacheManager.caching({
    store: 'memory',
    ttl: 60 * 60 // 1 hour
  })
}
