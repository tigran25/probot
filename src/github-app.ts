import jwt from 'jsonwebtoken'

export const createApp = (options: AppOptions) => {
  return () => {
    const payload = {
      exp: Math.floor(Date.now() / 1000) + 60,  // JWT expiration time
      iat: Math.floor(Date.now() / 1000),       // Issued at time
      iss: options.id                           // GitHub App ID
    }

    // Sign with RSA SHA256
    return jwt.sign(payload, options.cert, {algorithm: 'RS256'})
  }
}

export interface AppOptions {
  id: number
  cert: string
}
