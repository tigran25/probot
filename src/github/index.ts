import * as Octokit from '@octokit/rest'
import {addGraphQL} from './graphql'
import {addLogging, Logger} from './logging'
import {addPagination} from './pagination'
import {addRateLimiting} from './rate-limiting'

/**
 * the [@octokit/rest Node.js module](https://github.com/octokit/rest.js),
 * which wraps the [GitHub API](https://developer.github.com/v3/) and allows
 * you to do almost anything programmatically that you can do through a web
 * browser.
 * @typedef github
 * @see {@link https://github.com/octokit/rest.js}
 */
export function GitHubAPI(options: Options = {} as any) {
  const octokit = new Octokit(options) as GitHubAPI

  addRateLimiting(octokit, options.limiter)
  addLogging(octokit, options.logger)
  addPagination(octokit)
  addGraphQL(octokit)

  return octokit
}

export interface Options extends Octokit.Options {
  debug: boolean
  logger: Logger
  limiter?: any
}

export interface RequestOptions {
  method: string
  url: string
  headers: any
  query?: string
  variables?: Variables
}

export interface Result {
  headers: {
    status: string
  }
}

export interface OctokitError {
  code: number
  status: string
}

export interface GitHubAPI extends Octokit {
  paginate: (res: Promise<Octokit.AnyResponse>, callback: (results: Octokit.AnyResponse) => void) => Promise<any[]>
  // The following are added because Octokit does not expose the hook.error, hook.before, and hook.after methods
  hook: {
    error: (when: 'request', callback: (error: OctokitError, options: RequestOptions) => void) => void
    before: (when: 'request', callback: (result: Result, options: RequestOptions) => void) => void
    after: (when: 'request', callback: (result: Result, options: RequestOptions) => void) => void
  }

  request: (RequestOptions) => Promise<Octokit.AnyResponse>
  query: (query: string, variables?: Variables, headers?: Headers) => Promise<Octokit.AnyResponse>
}

export interface Headers {
  [key: string]: string
}

export interface Variables { [key: string]: any }
