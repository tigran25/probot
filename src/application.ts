import express from 'express'
import { EventEmitter } from 'promise-events'
import { ApplicationFunction } from '.'
import { Context } from './context'
import { GitHubAPI } from './github'
import { logger } from './logger'
import { LoggerWithTarget, wrapLogger } from './wrap-logger'

// Some events can't get an authenticated client (#382):
function isUnauthenticatedEvent (context: Context) {
  return !context.payload.installation ||
    (context.event === 'installation' && context.payload.action === 'deleted')
}

/**
 * The `app` parameter available to apps
 *
 * @property {logger} log - A logger
 */
export class Application {
  public events: EventEmitter
  public app: () => string
  public cache: Cache
  public router: express.Router
  public catchErrors: boolean
  public log: LoggerWithTarget

  constructor (options?: Options) {
    const opts = options || {} as any
    this.events = new EventEmitter()
    this.log = wrapLogger(logger, logger)
    this.app = opts.app
    this.cache = opts.cache
    this.catchErrors = opts.catchErrors || false
    this.router = opts.router || express.Router() // you can do this?
  }

  /**
   * Loads a Probot plugin
   * @param plugin - Probot plugin to load
   */
  public load (app: ApplicationFunction | ApplicationFunction[]): Application {
    if (Array.isArray(app)) {
      app.forEach(a => this.load(a))
    } else {
      app(this)
    }

    return this
  }

  public async receive (event: WebhookEvent) {
    return Promise.all([
      this.events.emit('*', event),
      this.events.emit(event.event, event),
      this.events.emit(`${event.event}.${event.payload.action}`, event)
    ])
  }

  /**
   * Get an {@link http://expressjs.com|express} router that can be used to
   * expose HTTP endpoints
   *
   * ```
   * module.exports = app => {
   *   // Get an express router to expose new HTTP endpoints
   *   const route = app.route('/my-app');
   *
   *   // Use any middleware
   *   route.use(require('express').static(__dirname + '/public'));
   *
   *   // Add a new route
   *   route.get('/hello-world', (req, res) => {
   *     res.end('Hello World');
   *   });
   * };
   * ```
   *
   * @param path - the prefix for the routes
   * @returns an [express.Router](http://expressjs.com/en/4x/api.html#router)
   */
  public route (path?: string): express.Router {
    if (path) {
      const router = express.Router()
      this.router.use(path, router)
      return router
    } else {
      return this.router
    }
  }

  /**
   * Listen for [GitHub webhooks](https://developer.github.com/webhooks/),
   * which are fired for almost every significant action that users take on
   * GitHub.
   *
   * @param event - the name of the [GitHub webhook
   * event](https://developer.github.com/webhooks/#events). Most events also
   * include an "action". For example, the * [`issues`](
   * https://developer.github.com/v3/activity/events/types/#issuesevent)
   * event has actions of `assigned`, `unassigned`, `labeled`, `unlabeled`,
   * `opened`, `edited`, `milestoned`, `demilestoned`, `closed`, and `reopened`.
   * Often, your bot will only care about one type of action, so you can append
   * it to the event name with a `.`, like `issues.closed`.
   *
   * ```js
   * app.on('push', context => {
   *   // Code was just pushed.
   * });
   *
   * app.on('issues.opened', context => {
   *   // An issue was just opened.
   * });
   * ```
   *
   * @param callback - a function to call when the
   * webhook is received.
   */
  public on (eventName: string | string[], callback: (context: Context) => Promise<void>) {
    if (typeof eventName === 'string') {

      return this.events.on(eventName, async (event: Context) => {
        const log = this.log.child({ name: 'event', id: event.id })

        try {
          let github

          if (isUnauthenticatedEvent(event)) {
            github = await this.auth()
            log.debug('`context.github` is unauthenticated. See https://probot.github.io/docs/github-api/#unauthenticated-events')
          } else {
            github = await this.auth(event.payload.installation.id, log)
          }

          const context = new Context(event, github, log)

          await callback(context)
        } catch (err) {
          log.error({ err, event })
          if (!this.catchErrors) {
            throw err
          }
        }
      })
    } else {
      eventName.forEach(e => this.on(e, callback))
    }
  }

  /**
   * Authenticate and get a GitHub client that can be used to make API calls.
   *
   * You'll probably want to use `context.github` instead.
   *
   * **Note**: `app.auth` is asynchronous, so it needs to be prefixed with a
   * [`await`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/await)
   * to wait for the magic to happen.
   *
   * ```js
   *  module.exports = (app) => {
   *    app.on('issues.opened', async context => {
   *      const github = await app.auth();
   *    });
   *  };
   * ```
   *
   * @param id - ID of the installation, which can be extracted from
   * `context.payload.installation.id`. If called without this parameter, the
   * client wil authenticate [as the app](https://developer.github.com/apps/building-integrations/setting-up-and-registering-github-apps/about-authentication-options-for-github-apps/#authenticating-as-a-github-app)
   * instead of as a specific installation, which means it can only be used for
   * [app APIs](https://developer.github.com/v3/apps/).
   *
   * @returns An authenticated GitHub API client
   * @private
   */
  public async auth (id?: number, log = this.log): Promise<GitHubAPI> {
    if (process.env.GHE_HOST && /^https?:\/\//.test(process.env.GHE_HOST)) {
      throw new Error('Your \`GHE_HOST\` environment variable should not begin with https:// or http://')
    }

    const github = GitHubAPI({
      baseUrl: process.env.GHE_HOST && `https://${process.env.GHE_HOST}/api/v3`,
      debug: process.env.LOG_LEVEL === 'trace',
      logger: log.child({ name: 'github', installation: String(id) })
    })

    if (id) {
      const res = await this.cache.wrap(`app:${id}:token`, () => {
        log.trace(`creating token for installation`)
        github.authenticate({ type: 'app', token: this.app() })

        return github.apps.createInstallationToken({ installation_id: String(id) })
      }, { ttl: 60 * 59 }) // Cache for 1 minute less than GitHub expiry

      github.authenticate({ type: 'token', token: res.data.token })
    } else {
      github.authenticate({ type: 'app', token: this.app() })
    }

    return github
  }
}

export interface WebhookEvent {
  event: string
  id: string
  payload: any
  protocol: 'http' | 'https'
  host: string
  url: string
}

// The TypeScript definition for cache-manager does not export the Cache interface so we recreate it here
export interface Cache {
  wrap<T> (key: string, wrapper: (callback: (error: any, result: T) => void) => any, options: CacheConfig): Promise<any>
}
export interface CacheConfig {
  ttl: number
}

export interface Options {
  app: () => string
  cache: Cache
  router?: express.Router
  catchErrors?: boolean
}
