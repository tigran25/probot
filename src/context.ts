import Webhooks, { PayloadRepository } from '@octokit/webhooks'
import yaml from 'js-yaml'
import path from 'path'
import { GitHubAPI } from './github'
import { LoggerWithTarget } from './wrap-logger'

interface WebhookPayloadWithRepository {
  [key: string]: any
  repository?: PayloadRepository
  issue?: {
    [key: string]: any
    number: number
    html_url?: string
    body?: string
  }
  pull_request?: {
    [key: string]: any
    number: number
    html_url?: string
    body?: string
  }
  sender?: {
    [key: string]: any
    type: string
  }
  action?: string
  installation?: {
    id: number
    [key: string]: any
  }
}

/**
 * The context of the event that was triggered, including the payload and
 * helpers for extracting information can be passed to GitHub API calls.
 *
 *  ```js
 *  module.exports = app => {
 *    app.on('push', context => {
 *      context.log('Code was pushed to the repo, what should we do with it?');
 *    });
 *  };
 *  ```
 *
 * @property {github} github - A GitHub API client
 * @property {payload} payload - The webhook event payload
 * @property {logger} log - A logger
 */

export class Context<E extends WebhookPayloadWithRepository = any> implements Webhooks.WebhookEvent<E> {
  public name: string
  public id: string
  public payload: E
  public protocol?: 'http' | 'https'
  public host?: string
  public url?: string

  public github: GitHubAPI
  public log: LoggerWithTarget

  constructor (event: Webhooks.WebhookEvent<E>, github: GitHubAPI, log: LoggerWithTarget) {
    this.name = event.name
    this.id = event.id
    this.payload = event.payload
    this.protocol = event.protocol
    this.host = event.host
    this.url = event.url

    this.github = github
    this.log = log
  }

  // Maintain backward compatibility
  public get event (): string {
    return this.name
  }

  /**
   * Return the `owner` and `repo` params for making API requests against a
   * repository.
   *
   * ```js
   * const params = context.repo({path: '.github/config.yml'})
   * // Returns: {owner: 'username', repo: 'reponame', path: '.github/config.yml'}
   * ```
   *
   * @param object - Params to be merged with the repo params.
   *
   */
  public repo<T> (object?: T) {
    const repo = this.payload.repository

    if (!repo) {
      throw new Error('context.repo() is not supported for this webhook event.')
    }

    return Object.assign({
      owner: repo.owner.login || repo.owner.name!,
      repo: repo.name
    }, object)
  }

  /**
   * Return the `owner`, `repo`, and `number` params for making API requests
   * against an issue or pull request. The object passed in will be merged with
   * the repo params.
   *
   * ```js
   * const params = context.issue({body: 'Hello World!'})
   * // Returns: {owner: 'username', repo: 'reponame', number: 123, body: 'Hello World!'}
   * ```
   *
   * @param object - Params to be merged with the issue params.
   */
  public issue<T> (object?: T) {
    const payload = this.payload
    return Object.assign({
      number: (payload.issue || payload.pull_request || payload).number
    }, this.repo(object))
  }

  /**
   * Returns a boolean if the actor on the event was a bot.
   * @type {boolean}
   */
  get isBot () {
    return this.payload.sender!.type === 'Bot'
  }

  /**
   * Reads the app configuration from the given YAML file in the `.github`
   * directory of the repository.
   *
   * For example, given a file named `.github/config.yml`:
   *
   * ```yml
   * close: true
   * comment: Check the specs on the rotary girder.
   * ```
   *
   * Your app can read that file from the target repository:
   *
   * ```js
   * // Load config from .github/config.yml in the repository
   * const config = await context.config('config.yml')
   *
   * if (config.close) {
   *   context.github.issues.comment(context.issue({body: config.comment}))
   *   context.github.issues.edit(context.issue({state: 'closed'}))
   * }
   * ```
   *
   * You can also use a `defaultConfig` object:
   *
   * ```js
   * // Load config from .github/config.yml in the repository and combine with default config
   * const config = await context.config('config.yml', {comment: 'Make sure to check all the specs.'})
   *
   * if (config.close) {
   *   context.github.issues.comment(context.issue({body: config.comment}));
   *   context.github.issues.edit(context.issue({state: 'closed'}))
   * }
   * ```
   *
   * @param fileName - Name of the YAML file in the `.github` directory
   * @param defaultConfig - An object of default config options
   * @return Configuration object read from the file
   */
  public async config<T> (fileName: string, defaultConfig?: T) {
    const params = this.repo({ path: path.posix.join('.github', fileName) })

    try {
      const res = await this.github.repos.getContents(params)
      const config = yaml.safeLoad(Buffer.from(res.data.content, 'base64').toString()) || {}
      return Object.assign({}, defaultConfig, config)
    } catch (err) {
      if (err.status === 404) {
        if (defaultConfig) {
          return defaultConfig
        }
        return null
      } else {
        throw err
      }
    }
  }
}
