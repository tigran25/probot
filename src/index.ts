import * as Logger from 'bunyan'
import * as cacheManager from 'cache-manager'
import * as express from 'express'
import {Application, WebhookEvent} from './application'
import {Context} from './context'
import {createApp} from './github-app'
import {logger} from './logger'
import {resolve} from './resolver'
import {createServer} from './server'
import {createWebhookProxy} from './webhook-proxy'

const Webhooks = require('@octokit/webhooks')
const logRequestErrors = require('./middleware/log-request-errors')

const cache = cacheManager.caching({
  store: 'memory',
  ttl: 60 * 60 // 1 hour
})

const defaultApps = [
  require('./plugins/sentry'),
  require('./plugins/stats'),
  require('./plugins/default')
]

export class Probot {
  public server: express.Application
  public webhook: any
  public logger: Logger

  private options: Options
  private apps: Application[]
  private app: () => string

  constructor(options: Options) {
    options.webhookPath = options.webhookPath || '/'
    options.secret = options.secret || 'development'
    this.options = options
    this.logger = logger
    this.apps = []
    this.webhook = new Webhooks({path: options.webhookPath, secret: options.secret})
    this.app = createApp({ id: options.id, cert: options.cert })
    this.server = createServer({webhook: this.webhook.middleware, logger})

    // Log all received webhooks
    this.webhook.on('*', (event: any) => {
      const webhookEvent = { ...event, event: event.name }
      delete webhookEvent.name

      this.receive(webhookEvent)
    })

    // Log all webhook errors
    this.webhook.on('error', this.errorHandler)
  }

  public errorHandler (err) {
    switch (err.message) {
      case 'X-Hub-Signature does not match blob signature':
      case 'No X-Hub-Signature found on request':
        logger.error('Go to https://github.com/settings/apps/YOUR_APP and verify that the Webhook secret matches the value of the WEBHOOK_SECRET environment variable.')
        break
      case 'error:0906D06C:PEM routines:PEM_read_bio:no start line':
      case '{"message":"A JSON web token could not be decoded","documentation_url":"https://developer.github.com/v3"}':
        logger.error('Your private key (usually a .pem file) is not correct. Go to https://github.com/settings/apps/YOUR_APP and generate a new PEM file. If you\'re deploying to Now, visit https://probot.github.io/docs/deployment/#now.')
        break
      default:
        logger.error(err)
    }
  }

  public receive (event: WebhookEvent) {
    this.logger.debug({event}, 'Webhook received')
    return Promise.all(this.apps.map(app => app.receive(event)))
  }

  public load (plugin: string | Plugin) {
    if (typeof plugin === 'string') {
      plugin = resolve(plugin) as Plugin
    }

    const app = new Application({app: this.app, cache, catchErrors: true})

    // Connect the router from the app to the server
    this.server.use(app.router)

    // Initialize the plugin
    plugin(app)
    this.apps.push(app)

    return app
  }

  public setup (apps: Array<string | Plugin>) {
    // Log all unhandled rejections
    process.on('unhandledRejection', this.errorHandler)

    // Load the given apps along with the default apps
    apps.concat(defaultApps).forEach(app => this.load(app))

    // Register error handler as the last middleware
    this.server.use(logRequestErrors)
  }

  public start () {
    if (this.options.webhookProxy) {
      createWebhookProxy({
        logger,
        path: this.options.webhookPath,
        port: this.options.port,
        url: this.options.webhookProxy,
      })
    }

    this.server.listen(this.options.port)
    logger.info('Listening on http://localhost:' + this.options.port)
  }
}

export const createProbot = (options: Options) => new Probot(options)

export type Plugin = (app: Application) => void

export interface Options {
  webhookPath?: string
  secret?: string,
  id: number,
  cert: string,
  webhookProxy?: string,
  port?: number
}

export { Logger, Context, Application }
