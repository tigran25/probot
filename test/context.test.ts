const fs = require('fs')
const path = require('path')

import {Context} from '../src/context'
import {OctokitError} from '../src/github'

describe('Context', () => {
  let event
  let context

  beforeEach(() => {
    event = {
      event: 'push',
      payload: {
        issue: {number: 4},
        repository: {
          name: 'probot',
          owner: {login: 'bkeepers'},
        },
      }
    }

    context = new Context(event, {} as any, {} as any)
  })

  it('inherits the payload', () => {
    expect(context.payload).toBe(event.payload)
  })

  describe('repo', () => {
    it('returns attributes from repository payload', () => {
      expect(context.repo()).toEqual({owner: 'bkeepers', repo: 'probot'})
    })

    it('merges attributes', () => {
      expect(context.repo({foo: 1, bar: 2})).toEqual({
        bar: 2, foo: 1, owner: 'bkeepers', repo: 'probot',
      })
    })

    it('overrides repo attributes', () => {
      expect(context.repo({owner: 'muahaha'})).toEqual({
        owner: 'muahaha', repo: 'probot'
      })
    })

    // The `repository` object on the push event has a different format than the other events
    // https://developer.github.com/v3/activity/events/types/#pushevent
    it('properly handles the push event', () => {
      event.payload = require('./fixtures/webhook/push')

      context = new Context(event, {} as any, {} as any)
      expect(context.repo()).toEqual({owner: 'bkeepers-inc', repo: 'test'})
    })

    it('return error for context.repo() when repository doesn\'t exist', () => {
      delete context.payload.repository
      try {
        context.repo()
      } catch (e) {
        expect(e.message).toMatch('context.repo() is not supported')
      }
    })
  })

  describe('issue', () => {
    it('returns attributes from repository payload', () => {
      expect(context.issue()).toEqual({owner: 'bkeepers', repo: 'probot', number: 4})
    })

    it('merges attributes', () => {
      expect(context.issue({foo: 1, bar: 2})).toEqual({
        bar: 2, foo: 1, number: 4, owner: 'bkeepers', repo: 'probot',
      })
    })

    it('overrides repo attributes', () => {
      expect(context.issue({owner: 'muahaha', number: 5})).toEqual({
        number: 5, owner: 'muahaha', repo: 'probot'
      })
    })
  })

  describe('config', () => {
    let github

    function readConfig (fileName) {
      const configPath = path.join(__dirname, 'fixtures', 'config', fileName)
      const content = fs.readFileSync(configPath, {encoding: 'utf8'})
      return {data: {content: Buffer.from(content).toString('base64')}}
    }

    beforeEach(() => {
      github = {
        repos: {
          getContent: jest.fn()
        }
      }

      context = new Context(event, github, {} as any)
    })

    it('gets a valid configuration', async () => {
      github.repos.getContent = jest.fn().mockReturnValue(Promise.resolve(readConfig('basic.yml')))
      const config = await context.config('test-file.yml')

      expect(github.repos.getContent).toHaveBeenCalledWith({
        owner: 'bkeepers',
        path: '.github/test-file.yml',
        repo: 'probot',
      })
      expect(config).toEqual({
        bar: 7,
        baz: 11,
        foo: 5,
      })
    })

    it('returns null when the file is missing', async () => {
      const error: OctokitError = {
        code: 404,
        message: 'An error occurred',
        name: 'OctokitError',
        status: 'Not Found',
      };

      github.repos.getContent = jest.fn().mockReturnValue(Promise.reject(error))

      expect(await context.config('test-file.yml')).toBe(null)
    })

    it('returns the default config when the file is missing and default config is passed', async () => {
      const error: OctokitError = {
        code: 404,
        message: 'An error occurred',
        name: 'OctokitError',
        status: 'Not Found',
      };

      github.repos.getContent = jest.fn().mockReturnValue(Promise.reject(error))
      const defaultConfig = {
        bar: 7,
        baz: 11,
        foo: 5,
      }
      const contents = await context.config('test-file.yml', defaultConfig)
      expect(contents).toEqual(defaultConfig)
    })

    it('throws when the configuration file is malformed', async () => {
      github.repos.getContent = jest.fn().mockReturnValue(Promise.resolve(readConfig('malformed.yml')))

      let e
      let contents
      try {
        contents = await context.config('test-file.yml')
      } catch (err) {
        e = err
      }

      expect(contents).toBeUndefined()
      expect(e).toBeDefined()
      expect(e.message).toMatch(/^end of the stream or a document separator/)
    })

    it('throws when loading unsafe yaml', async () => {
      github.repos.getContent = jest.fn().mockReturnValue(readConfig('evil.yml'))

      let e
      let config
      try {
        config = await context.config('evil.yml')
      } catch (err) {
        e = err
      }

      expect(config).toBeUndefined()
      expect(e).toBeDefined()
      expect(e.message).toMatch(/unknown tag/)
    })

    it('returns an empty object when the file is empty', async () => {
      github.repos.getContent = jest.fn().mockReturnValue(readConfig('empty.yml'))

      const contents = await context.config('test-file.yml')

      expect(contents).toEqual({})
    })

    it('overwrites default config settings', async () => {
      github.repos.getContent = jest.fn().mockReturnValue(Promise.resolve(readConfig('basic.yml')))
      const config = await context.config('test-file.yml', {foo: 10})

      expect(github.repos.getContent).toHaveBeenCalledWith({
        owner: 'bkeepers',
        path: '.github/test-file.yml',
        repo: 'probot',
      })
      expect(config).toEqual({
        bar: 7,
        baz: 11,
        foo: 5,
      })
    })

    it('uses default settings to fill in missing options', async () => {
      github.repos.getContent = jest.fn().mockReturnValue(Promise.resolve(readConfig('missing.yml')))
      const config = await context.config('test-file.yml', {bar: 7})

      expect(github.repos.getContent).toHaveBeenCalledWith({
        owner: 'bkeepers',
        path: '.github/test-file.yml',
        repo: 'probot',
      })
      expect(config).toEqual({
        bar: 7,
        baz: 11,
        foo: 5,
      })
    })
  })
})
