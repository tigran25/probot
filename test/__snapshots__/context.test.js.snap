// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Context repo return error for context.repo() when repository doesn't exist 1`] = `"context.repo() is not supported for this webhook event."`;
